native/src/routePlannerFrontEnd.cpp

```diff
@@ -227,6 +227,448 @@ vector<SHARED_PTR<RouteSegmentResult>> RoutePlannerFrontEnd::searchRouteInternal
+			}
+		} else {
+			r->insert(pind, i->first, i->second);
+		}
+	}
+}
+
+bool stepBackAndFindPrevPointInRoute(GpxRouteApproximation* gctx, vector<SHARED_PTR<GpxPoint>> gpxPoints,
+									 SHARED_PTR<GpxPoint> start, SHARED_PTR<GpxPoint> next) {
+	// step back to find to be sure
+	// 1) route point is behind GpxPoint - MINIMUM_POINT_APPROXIMATION (end route point could slightly ahead)
+	// 2) we don't miss correct turn i.e. points could be attached to muliple routes
+	// 3) to make sure that we perfectly connect to RoadDataObject points
+	double STEP_BACK_DIST = max(gctx->MINIMUM_POINT_APPROXIMATION, gctx->MINIMUM_STEP_APPROXIMATION);
+	double d = 0;
+	int segmendInd = start->routeToTarget.size() - 1;
+	bool search = true;
+	start->stepBackRoute.clear();
+mainLoop:
+	for (; segmendInd >= 0 && search; segmendInd--) {
+		RouteSegmentResult rr = *start->routeToTarget[segmendInd];
+		bool minus = rr.getStartPointIndex() < rr.getEndPointIndex();
+		int nextInd;
+		for (int j = rr.getEndPointIndex(); j != rr.getStartPointIndex(); j = nextInd) {
+			nextInd = minus ? j - 1 : j + 1;
+			d += getDistance(rr.getPoint(j)->lat, rr.getPoint(j)->lon, rr.getPoint(nextInd)->lat,
+							 rr.getPoint(nextInd)->lon);
+			if (d > STEP_BACK_DIST) {
+				if (nextInd == rr.getStartPointIndex()) {
+					segmendInd--;
+				} else {
+					start->stepBackRoute.push_back(std::make_shared<RouteSegmentResult>(
+						RouteSegmentResult(rr.object, nextInd, rr.getEndPointIndex())));
+					rr.setEndPointIndex(nextInd);
+				}
+				search = false;
+				goto mainLoop;
+			}
+		}
+	}
+	if (segmendInd == -1) {
+		// here all route segments - 1 is longer than needed distance to step back
+		return false;
+	}
+
+	while (start->routeToTarget.size() > segmendInd + 1) {
+		SHARED_PTR<RouteSegmentResult> removed = start->routeToTarget[segmendInd + 1];
+		start->routeToTarget.erase(start->routeToTarget.begin() + segmendInd + 1);
+		start->stepBackRoute.push_back(removed);
+	}
+	SHARED_PTR<RouteSegmentResult> res = start->routeToTarget[segmendInd];
+	next->pnt = std::make_shared<RouteSegmentPoint>(RouteSegmentPoint(res->object, res->getEndPointIndex()));
+	return true;
+}
+
+void calculateGpxRoute(GpxRouteApproximation* gctx, vector<SHARED_PTR<GpxPoint>> gpxPoints) {
+	RoutingIndex* reg = new RoutingIndex();
+	reg->initRouteEncodingRule(0, "highway", UNMATCHED_HIGHWAY_TYPE);
+	vector<SHARED_PTR<LatLon>> lastStraightLine;
+	SHARED_PTR<GpxPoint> straightPointStart = NULL;
+	for (int i = 0; i < gpxPoints.size() && !gctx->ctx->progress->isCancelled();) {
+		SHARED_PTR<GpxPoint> pnt = gpxPoints[i];
+		if (!pnt->routeToTarget.empty()) {
+			SHARED_PTR<LatLon> startPoint = std::make_shared<LatLon>(pnt->routeToTarget[0]->getStartPoint()->lat,
+																	 pnt->routeToTarget[0]->getStartPoint()->lon);
+			if (!lastStraightLine.empty()) {
+				lastStraightLine.push_back(startPoint);
+				addStraightLine(gctx, lastStraightLine, straightPointStart, reg);
```

Comment:
suspicious to double check: this won't work if straightPointStart is set in addStraightLine method